### ISSUE TITLE
On Self Modulation for Generative Adversarial Networks

### ISSUE THREAD

**Issue by KeisukeShimokawa** [KeisukeShimokawa/papers-challenge#140]
### 論文へのリンク

[[arXiv:1810.01365] On Self Modulation for Generative Adversarial Networks](https://arxiv.org/abs/1810.01365)

### 著者・所属機関

Ting Chen, Mario Lucic, Neil Houlsby, Sylvain Gelly

- University of California, Los Angeles
- Google Brain

### 投稿日時（YYYY-MM-DD）

2018-10-02

## 2. 先行研究と比べてどこがすごいの？

既存のGANでSOTAを達成しているモデルでは、CBN (Conditional Batch Normalizatin) が使用されているが、この手法ではラベルなどの外部情報が必要であり常に使用できるとは限らない。

本研究では外部情報を必要とせず、Generatorの出力のみを使用するSelf-Modulation層を提案した。本モジュールはどのGeneratorにも追加できる。

## 3. 技術や手法の"キモ"はどこにある？

外部情報を使用してGeneratorを改善する手法としては、(1)入力されるノイズベクトルにラベル情報を結合させるやり方と、(2)CBNを使用して隠れ層に直接ラベル情報を送るやり方が存在している。

まずはBNを復習すると、入力された特徴マップhlに対して平均と標準偏差を使用して正規化を行い、追加のパラメータでアフィン変換を行っている。

![image](https://user-images.githubusercontent.com/33191285/83961199-cda21380-a8cb-11ea-8772-b12805579426.png)

### 3.1 Self-modulation for unconditional (without side information) generation

本研究で提案しているSelf-Modulationでは、最初に入力されたノイズベクトルzを使用してアフィン変換で使用していたγとβを条件付けを行っている。

![image](https://user-images.githubusercontent.com/33191285/83961286-8ff1ba80-a8cc-11ea-9835-f2d2d76b025d.png)

各パラメータの計算には、入力されるノイズベクトルzに対して1層の次元数の少ないMLP層とReLUを適用し、その後に線形変換を施している。

![image](https://user-images.githubusercontent.com/33191285/83961330-e8c15300-a8cc-11ea-88cc-e9b8efda393c.png)

### 3.2 Self-modulation for conditional (with side information) generation

ラベル情報を使用できる場合にも本手法にその情報を組み込むことが可能である。ラベル情報とノイズベクトルに関して2つのEmbedding層を使用してラベルベクトルに変換した後で、これら2つのベクトルとの相互関係を計算している。

![image](https://user-images.githubusercontent.com/33191285/83961429-e6132d80-a8cd-11ea-9c26-da9b392aa47b.png)

以下が各Normalization手法との比較である。

![image](https://user-images.githubusercontent.com/33191285/83961457-28d50580-a8ce-11ea-841d-9e24b38582fe.png)

## 4. どうやって有効だと検証した？

以下がBaselineと本手法を適用した際の比較である。

![image](https://user-images.githubusercontent.com/33191285/83961575-72722000-a8cf-11ea-92d1-ecb7a39087ae.png)

**Comment by KeisukeShimokawa** [comment]
```python
import torch.nn as nn

class sBN(nn.Module):
    def __init__(self, num_channels, nz, nz_hidden_channels):
        super().__init__()
        self.num_channels = num_channels

        self.bn = nn.BatchNorm2d(num_channels, affine=False)
        self.z_to_h = nn.Linear(nz, nz_hidden_channels)
        self.h_to_gamma = nn.Linear(nz_hidden_channels, num_channels)
        self.h_to_beta = nn.Linear(nz_hidden_channels, num_channels)
        self.activation = nn.ReLU(inplace=True)

    def forward(self, x, z):
        x_norm = self.bn(x)
        h = self.activation(self.z_to_h(z))
        gamma = self.h_to_gamma(h).view(-1, self.num_channels, 1, 1)
        beta = self.h_to_beta(h).view(-1, self.num_channels, 1, 1)
        output = gamma * x_norm + beta
        return output
```
